Changes for Hibernate Reactive

DavideD opened the pull request:
Fix [HHH-18854](https://hibernate.atlassian.net/browse/HHH-18854) (I think)

These changes make it possible for me to use Hibernate Reactive with ORM 7 and Postgres.
I still need to fix some issues with the other databases, but I don't think I will have more changes to include for now.

Basically, I needed some way to intercept the creation of the id generator.

[HHH-18854]: https://hibernate.atlassian.net/browse/HHH-18854?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

gavinking reviewed (commented):
@DavideD Could you please add comments to every new method and constructor explaining that it's used by Reactive.

DavideD commented:
Comments added

gavinking commented:
> Comments added

Thanks man

DavideD commented:
Rebased to the latest changes and added a minor refactoring: https://github.com/hibernate/hibernate-orm/pull/9264/commits/f0b8dca0cb0f8be6fcc2ecb1aff64a17d7b9814e

gavinking commented on hibernate-core/src/main/java/org/hibernate/persister/entity/AbstractEntityPersister.java:
```diff
@@ -469,11 +469,38 @@ public abstract class AbstractEntityPersister
 	private List<UniqueKeyEntry> uniqueKeyEntries = null; //lazily initialized
 	private ConcurrentHashMap<String,SingleIdArrayLoadPlan> nonLazyPropertyLoadPlansByName;
 
+	/**
+	 * A factory for the creation of an EntityMetamodel.
+	 * <p>
+	 * Hibernate Reactive uses it to pass its own entity metamodel class
+	 * and adapt the identifier generators.
+	 */
+	public static class EntityMetamodelFactory {
+		public EntityMetamodel createEntityMetamodel(
+				PersistentClass persistentClass,
+				EntityPersister persister,
+				RuntimeModelCreationContext creationContext) {
+			return new EntityMetamodel( persistentClass, persister, creationContext );
+		}
+	}
```
@DavideD instead of introducing this new class, could you just add a new `createEntityMetamodel()` operation to `RuntimeModelCreationContext`?

Or, if that doesn't work, could you at least make it a `@FunctionInterface`?

DavideD replied:
I will check

DavideD commented:
@sebersole Is it OK to merge this before the next release?

DavideD commented:
@gavinking Is this better? https://github.com/hibernate/hibernate-orm/pull/9264/commits/7be8d0b27dd10e0879849c4279324aa1d66956b2

gavinking commented:
> @gavinking Is this better? [7be8d0b](https://github.com/hibernate/hibernate-orm/commit/7be8d0b27dd10e0879849c4279324aa1d66956b2)

Um, well, it's less code, so ... I guess??

gavinking commented on hibernate-core/src/main/java/org/hibernate/metamodel/spi/RuntimeModelCreationContext.java:
```diff
@@ -64,4 +67,11 @@ default MetadataImplementor getMetadata() {
+	/*
+	 * Used by Hibernate Reactive
+	 */
+	default EntityMetamodel createEntityMetamodel(PersistentClass persistentClass, EntityPersister persister, RuntimeModelCreationContext creationContext) {
+		return new EntityMetamodel( persistentClass, persister, creationContext );
+	}
```
```suggestion
	default EntityMetamodel createEntityMetamodel(PersistentClass persistentClass, EntityPersister persister) {
		return new EntityMetamodel( persistentClass, persister, this );
	}
```

Can just use `this` here.

DavideD replied:
Good catch, solved

sebersole commented:
> @sebersole Is it OK to merge this before the next release? (After I've applied the change Gavin asked)

Looks fine to me, I'll get it applied today.

sebersole commented:
We need to figure out this failure first though

sebersole commented:
https://github.com/hibernate/hibernate-orm/pull/9329

DavideD commented:
Thanks!
